File: Program/Testing Program/Test_Program_AGV/Serial_Com.cpp
/*  
  SERIAL COMMUNICATION LIBRARY
  PT. Stechoq Robotika Indonesia
  
  Date    : 27 SEPTEMBER 2024
  Author  : Alfonsus Giovanni Mahendra Putra - Universitas Diponegoro
*/

#include "Serial_Com.h"

static uint8_t
rx_buff[MAX_LEN];

uint8_t Length_calculator(uint8_t *data){
  uint8_t data_len = 0;
  for(size_t i=3; i<sizeof(data); i++){
    if(data[i] & data[i+1] != 0x00){
      data_len++;
    }
  }
  return data_len;
}

Com_status_t Ping_Return(Param_t *param){
  param->Error_value = 0;
  
  uint8_t 
  len = 0x02,
  tx_buff[] = {HEADER, HEADER, len, param->error_state.No_err, param->instruction.Ping};

  if(Serial.write(tx_buff, sizeof(tx_buff)) != 0) return Send_success;
  else return Serial_error;
}

Com_status_t Running_Mode_Return(Param_t *param){
  param->Error_value = 0;
  
  uint8_t 
  len = 0x03,
  tx_buff[] = {HEADER, HEADER, len, param->error_state.No_err, param->Select_mode, param->Base_speed};

  if(Serial.write(tx_buff, sizeof(tx_buff)) != 0) return Send_success;
  else return Serial_error;
}

Com_status_t Running_State_Return(Param_t *param){
  param->Error_value = 0;
  
  uint8_t 
  len = 0x04,
  tx_buff[] = {HEADER, HEADER, len, param->error_state.No_err, param->Select_state, param->Set_Acceleration, param->Set_Breaking};

  if(Serial.write(tx_buff, sizeof(tx_buff)) != 0) return Send_success;
  else return Serial_error;
}
Com_status_t AGV_Status_Return(Param_t *param){
  param->Error_value = 0;
  
  uint8_t 
  len = 0x0B,
  tx_buff[] = {
    HEADER, 
    HEADER, 
    len, 
    param->error_state.No_err, 
    param->Position,
    (param->Pos_value >> 8) & 0xFF,
    param->Pos_value & 0xFF,
    (param->Send_counter >> 8) & 0xFF,
    param->Send_counter & 0xFF,
    (param->Pickup_counter >> 8) & 0xFF,
    param->Pickup_counter & 0xFF,
    0x00,
    0x00,
    0x00,
    0x00
  };

  uint8_t byte_value[4];
  memcpy(byte_value, &param->Battery_level, sizeof(float));
  for(int i=0; i<4; i++) tx_buff[i+11] = byte_value[i]; 

  if(Serial.write(tx_buff, sizeof(tx_buff)) != 0) return Send_success;
  else return Serial_error;
}
Com_status_t Sensor_Data_Return(Param_t *param){
  param->Error_value = 0;
  
  uint8_t 
  len = 0x03,
  tx_buff[] = {HEADER, HEADER, len, param->error_state.No_err, param->SensorA, param->SensorB};

  if(Serial.write(tx_buff, sizeof(tx_buff)) != 0) return Send_success;
  else return Serial_error;
}
Com_status_t NFC_Data_Return(Param_t *param){
  param->Error_value = 0;
  
  uint8_t 
  len = 0x04,
  tx_buff[] = {
    HEADER, 
    HEADER, 
    len, 
    param->error_state.No_err,
    param->Tag_position, 
    (param->Tag_value >> 8) & 0xFF,
    param->Tag_value & 0xFF,
  };

  if(Serial.write(tx_buff, sizeof(tx_buff)) != 0) return Send_success;
  else return Serial_error;
}

Com_status_t Error_Packet_Return(Param_t *param){
  uint8_t
  len = 0x01,
  tx_buff[] = {HEADER, HEADER, len, param->Error_value};

  if(Serial.write(tx_buff, sizeof(tx_buff)) != 0) return Send_success;
  else return Serial_error;
}

Com_status_t Send_Ping(Param_t *param){

}
Com_status_t Send_Error(Param_t *param, uint8_t error_value){

}
Com_status_t Set_Running_Mode(Param_t *param, Select_mode_data_t mode, uint8_t base_speed){

}
Com_status_t Set_Running_State(Param_t *param, Select_state_data_t state, Accel_data_t accel, Brake_data_t brake){

}
Com_status_t Read_AGV_Status(Param_t *param, Position_data_t pos, uint16_t pos_val, uint16_t send_cnt, uint16_t pickup_cnt, float *battery_lvl){

}
Com_status_t Read_Sensor_Data(Param_t *param, Sensor_data_t sens_state){

}
Com_status_t Read_NFC_Data(Param_t *param, Position_data_t pos, uint8_t pos_val){
  
}

Com_status_t Receive_Instruction(Param_t *param){
  uint8_t data_len;
  uint8_t len_validation;

  // CHECK SERIAL AVAILABILITY
  if(Serial.available()){
    Serial.readBytes(rx_buff, MAX_LEN);
    len_validation = Length_calculator(rx_buff);

    if(rx_buff[0] == HEADER && rx_buff[1] == HEADER){
      data_len = rx_buff[2];
      // DATA LENGTH VALIDATION TRUE
      if(data_len == len_validation){
        if(rx_buff[3] == param->instruction.Ping){
          Ping_Return(param);
        }

        else if(rx_buff[3] == param->instruction.Read){
          if(rx_buff[4] == param->item.Running_Mode) Running_Mode_Return(param);
          else if(rx_buff[4] == param->item.Running_State) Running_State_Return(param);
          else if(rx_buff[4] == param->item.AGV_Status) AGV_Status_Return(param);
          else if(rx_buff[4] == param->item.Sensor_Data) Sensor_Data_Return(param);
          else if(rx_buff[4] == param->item.NFC_Data) NFC_Data_Return(param);
        }

        else if(rx_buff[3] == param->instruction.Write){
          if(rx_buff[4] == param->item.Running_Mode){
            param->Select_mode = rx_buff[5];
            param->Base_speed = rx_buff[6];

            param->Error_value = 0;
            Error_Packet_Return(param);
          }
          else if(rx_buff[4] == param->item.Running_State){
            param->Select_state = rx_buff[5];
            param->Set_Acceleration = rx_buff[6];
            param->Set_Breaking = rx_buff[7];

            param->Error_value = 0;
            Error_Packet_Return(param);
          }
          else if(rx_buff[4] == param->item.AGV_Status){
            uint8_t byte_data[4];

            param->Position = rx_buff[5];
            param->Pos_value = (rx_buff[6] << 8) | rx_buff[7];
            param->Send_counter = (rx_buff[8] << 8) | rx_buff[9];
            param->Pickup_counter = (rx_buff[10] << 8) | rx_buff[11];

            for(int i=0; i<4; i++) byte_data[i] = rx_buff[12+i];
            memcpy(&param->Battery_level, byte_data, sizeof(float));

            param->Error_value = 0;
            Error_Packet_Return(param);
          }
          else if(rx_buff[4] == param->item.Sensor_Data){
            param->SensorA = rx_buff[5];
            param->SensorB = rx_buff[6];

            param->Error_value = 0;
            Error_Packet_Return(param);
          }
          else if(rx_buff[4] == param->item.NFC_Data){
            param->Tag_position = rx_buff[5];
            param->Tag_value = (rx_buff[6] >> 8) | rx_buff[7];

            param->Error_value = 0;
            Error_Packet_Return(param);
          }

          // DATA ITEM INCORRECT
          else{
            param->Error_value = param->Error_value | param->error_state.Item_err;
            Error_Packet_Return(param);
          }
        }

        // DATA INSTRUCTION INCORRECT
        else{
          param->Error_value = param->Error_value | param->error_state.Instruction_err;
          Error_Packet_Return(param);
        }
      }

      // DATA LENGTH VALIDATION FALSE
      else{
        param->Error_value = param->Error_value | param->error_state.Length_err;
        Error_Packet_Return(param);
      }
    }
  }
}

File: Program/Raspberry Serial Com/SerialCom.py
# Raspberry Serial Communication Handler Program
# PT Stechoq Robotika Indonesia
# Date	 : 11 Oktober 2024
# Author : Alfonsus Giovanni Mahendra Putra - Universitas Diponegoro


# USED LIBRARY -----
import serial
import struct
import array as arr
import time
import copy
# ------------------

# SERIAL SETUP ---------
baudRate = 500000
MaxMsgsLen = 16
UARTPort = '/dev/serial0'
USBPort = '/dev/ttyACM1'
timeout = 1
# ----------------------

# Instruction 
class Instruction:
    Ping = 0x01
    Read = 0x02
    Write = 0x03

class Item:
    Running_Mode = 0x01
    Running_State = 0x02
    AGV_Status = 0x03
    Sensor_Data = 0x04
    NFC_Data = 0x05
    Joystick_Data = 0x06

class SubItem:
    Sub_item1 = 0x01
    Sub_item2 = 0x02
    Sub_item3 = 0x03

class Error:
    No_err = 0x00
    Instruction_err = 0x01
    Item_err = 0x02
    Length_err = 0x04
    Tag_reading_err = 0x08
    LoRa_com_err = 0x10
    Carrier_hook_err = 0x20
    Battery_overvoltage = 0x40
    Battery_overcurrent = 0x80

class SelectModeData:
    NOT_SET = 0x00
    LF_MODE = 0x01
    LIDAR_MODE = 0x02

class SelectSpeedData:
    NORMAL_SPEED = 0x01
    HIGH_SPEED = 0x02

class SelectStateData:
    START = 0x01
    STOP = 0x02
    PAUSE = 0x03

class SelectDirData:
    FORWARD = 0x01
    BACKWARD = 0x02
    LEFT = 0x03
    RIGHT = 0x04
    ROTATE_LEFT = 0x05
    ROTATE_RIGHT = 0x06
    BRAKE = 0x07

class SensorData:
    DETECTED = 0x01
    NOT_DETECTED = 0x02

class AccelData:
    NORMAL_ACCEL = 0x01
    REGENERATIVE_ACCEL = 0x02

class BrakeData:
    NORMAL_BRAKE = 0x01
    REGENERATIVE_BRAKE = 0x02

class PositionData:
    HOME = 0x01
    ON_STATION = 0x02
    ON_THE_WAY = 0x03


class Serial_COM:
    def __init__(self, inputBaudRate, serialPort, inputTimeout):
          self.baud = inputBaudRate
          self.port = serialPort
          self.timeout = inputTimeout
          
          self.Header = 0xFF

          self.return_data = bytearray(16)
          self.data_length = 0
          self.Instruction_get = 0
          self.Item_get = 0
          self.SubItem_get = 0
          
          self.Select_mode = SelectModeData.NOT_SET
          self.Base_speed = 0
          
          self.Select_state = SelectStateData.STOP
          self.Set_Direction = SelectDirData.FORWARD
          self.Set_Acceleration = AccelData.NORMAL_ACCEL
          self.Set_Breaking = BrakeData.NORMAL_BRAKE
          
          self.Position = PositionData.HOME
          self.Pos_value = 0
          self.Send_counter = 0
          self.Pickup_counter = 0
          self.Tag_position = PositionData.HOME
          self.Tag_value = 0
          
          self.SensorA = SensorData.NOT_DETECTED
          self.SensorB = SensorData.NOT_DETECTED
          
          self.Battery_level = 0.0
          
          self.Xpos = 0
          self.Ypos = 0
          
          self.instruction = Instruction()
          self.error_state = Error()
          self.item = Item()
          self.sub_item = SubItem()

          self.running_mode = SelectModeData()
          self.accel_mode = AccelData()
          self.brake_mode = BrakeData()
          
          self.data = serial.Serial(serialPort, inputBaudRate, timeout=inputTimeout)
          
    def Send_Ping(self):
        msgs_len = 0x01
        tx_buff = [self.Header, self.Header, msgs_len, self.instruction.Ping]
        self.data.write(tx_buff)

        rx_buff = self.data.read(5)
        if len(rx_buff) == 5:
            if rx_buff[0] == self.Header and rx_buff[1] == self.Header:
                self.data_length = rx_buff[2]
            self.return_data = copy.deepcopy(rx_buff)
            self.data.flushInput()
    
    def Set_Running_Mode(self, select_mode):
        msgs_len = 0x04
        tx_buff = [self.Header, self.Header, msgs_len, self.instruction.Write, self.item.Running_Mode, self.sub_item.Sub_item1, select_mode]
        self.data.write(tx_buff)

        rx_buff = self.data.read(4)
        if len(rx_buff) == 4:
            if rx_buff[0] == self.Header and rx_buff[1] == self.Header:
                self.data_length = rx_buff[2]
            self.return_data = copy.deepcopy(rx_buff)
            self.data.flushInput()
    
    def Set_Base_Speed(self, base_speed):
        msgs_len = 0x04
        tx_buff = [self.Header, self.Header, msgs_len, self.instruction.Write, self.item.Running_Mode, self.sub_item.Sub_item2, base_speed]
        self.data.write(tx_buff)

        rx_buff = self.data.read(4)
        if len(rx_buff) == 4:
            if rx_buff[0] == self.Header and rx_buff[1] == self.Header:
                self.data_length = rx_buff[2]
            self.return_data = copy.deepcopy(rx_buff)
            self.data.flushInput()

    def Set_Running_State(self, select_state, direction):
        msgs_len = 0x05
        tx_buff = [self.Header, self.Header, msgs_len, self.instruction.Write, self.item.Running_State, self.sub_item.Sub_item1, select_state, direction]
        self.data.write(tx_buff)

        rx_buff = self.data.read(4)
        if len(rx_buff) == 4:
            if rx_buff[0] == self.Header and rx_buff[1] == self.Header:
                self.data_length = rx_buff[2]
            self.return_data = copy.deepcopy(rx_buff)
            self.data.flushInput()
    
    def Set_Running_Accel(self, set_accel_mode):
        msgs_len = 0x04
        tx_buff = [self.Header, self.Header, msgs_len, self.instruction.Write, self.item.Running_State, self.sub_item.Sub_item2, set_accel_mode]
        self.data.write(tx_buff)

        rx_buff = self.data.read(4)
        if len(rx_buff) == 4:
            if rx_buff[0] == self.Header and rx_buff[1] == self.Header:
                self.data_length = rx_buff[2]
            self.return_data = copy.deepcopy(rx_buff)
            self.data.flushInput()
    
    def Set_Running_Brake(self, set_brake_mode):
        msgs_len = 0x04
        tx_buff = [self.Header, self.Header, msgs_len, self.instruction.Write, self.item.Running_State, self.sub_item.Sub_item3, set_brake_mode]
        self.data.write(tx_buff)

        rx_buff = self.data.read(4)
        if len(rx_buff) == 4:
            if rx_buff[0] == self.Header and rx_buff[1] == self.Header:
                self.data_length = rx_buff[2]
            self.return_data = copy.deepcopy(rx_buff)
            self.data.flushInput()
    
    def Set_Joystick(self, valueX, valueY):
        msgs_len = 0x04
        tx_buff = [self.Header, self.Header, msgs_len, self.instruction.Write, self.item.Running_State, valueX, valueY]
        self.data.write(tx_buff)

        rx_buff = self.data.read(4)
        if len(rx_buff) == 4:
            if rx_buff[0] == self.Header and rx_buff[1] == self.Header:
                self.data_length = rx_buff[2]
                self.return_data = copy.deepcopy(rx_buff)
                self.data.flushInput()
    
    def Read_Running_Mode(self):
        msgs_len = 0x02
        tx_buff = [self.Header, self.Header, msgs_len, self.instruction.Read, self.item.Running_Mode]
        self.data.write(tx_buff)

        rx_buff = self.data.read(6)
        if len(rx_buff) == 6:
            if rx_buff[0] == self.Header and rx_buff[1] == self.Header and rx_buff[3] == 0:
                self.data_length = rx_buff[2]

                if(rx_buff[4] == 0x00):
                    self.Select_mode = SelectModeData.NOT_SET
                elif(rx_buff[4] == 0x01):
                    self.Select_mode = SelectModeData.LF_MODE
                else:
                    self.Select_mode = SelectModeData.LIDAR_MODE
                self.Base_speed = rx_buff[5]

                self.return_data = copy.deepcopy(rx_buff)

            self.return_data = copy.deepcopy(rx_buff)
            self.data.flushInput()

    def Read_Running_State(self):
        msgs_len = 0x02
        tx_buff = [self.Header, self.Header, msgs_len, self.instruction.Read, self.item.Running_State]
        self.data.write(tx_buff)

        rx_buff = self.data.read(8)
        if len(rx_buff) == 8:
            if rx_buff[0] == self.Header and rx_buff[1] == self.Header and rx_buff[3] == 0:
                self.data_length = rx_buff[2]

                if(rx_buff[4] == 0x01):
                    self.Select_state = SelectStateData.START
                elif(rx_buff[4] == 0x02):
                    self.Select_state = SelectStateData.STOP
                else:
                    self.Select_state = SelectStateData.PAUSE

                if(rx_buff[5] == 0x01):
                    self.Set_Direction = SelectDirData.FORWARD
                elif(rx_buff[5] == 0x02):
                    self.Set_Direction = SelectDirData.BACKWARD
                elif(rx_buff[5] == 0x03):
                    self.Set_Direction = SelectDirData.LEFT
                elif(rx_buff[5] == 0x04):
                    self.Set_Direction = SelectDirData.RIGHT
                elif(rx_buff[5] == 0x05):
                    self.Set_Direction = SelectDirData.ROTATE_LEFT
                elif(rx_buff[5] == 0x06):
                    self.Set_Direction = SelectDirData.ROTATE_RIGHT
                elif(rx_buff[5] == 0x07):
                    self.Set_Direction = SelectDirData.BRAKE

                if(rx_buff[6] == 0x01):
                    self.Set_Acceleration = AccelData.NORMAL_ACCEL
                else:
                    self.Set_Acceleration = AccelData.REGENERATIVE_ACCEL

                if(rx_buff[7] == 0x01):
                    self.Set_Breaking = BrakeData.NORMAL_BRAKE
                else:
                    self.Set_Breaking = BrakeData.REGENERATIVE_BRAKE

                self.return_data = copy.deepcopy(8)

            self.return_data = copy.deepcopy(rx_buff)
            self.data.flushInput()
    
    def Read_AGV_Status(self):
        msgs_len = 0x02
        tx_buff = [self.Header, self.Header, msgs_len, self.instruction.Read, self.item.AGV_Status]
        self.data.write(tx_buff)

        rx_buff = self.data.read(15)
        if len(rx_buff) == 15:
            if rx_buff[0] == self.Header and rx_buff[1] == self.Header and rx_buff[3] == 0:
                self.data_length = rx_buff[2]

                if(rx_buff[4] == 0x01):
                    self.Position = PositionData.HOME
                elif(rx_buff[4] == 0x02):
                    self.Position = PositionData.ON_STATION
                else:
                    self.Position = PositionData.ON_THE_WAY

                self.Pos_value = (rx_buff[5] << 8) | rx_buff[6]
                self.Send_counter = (rx_buff[7] << 8) | rx_buff[8]
                self.Pickup_counter = (rx_buff[9] << 8) | rx_buff[10]

                voltage_array = arr.array('B', [rx_buff[11], rx_buff[12], rx_buff[13], rx_buff[14]])
                voltage_data = bytes(voltage_array)
                self.Battery_level = struct.unpack('f', voltage_data)[0]

                self.return_data = copy.deepcopy(rx_buff)

            self.return_data = copy.deepcopy(rx_buff)
            self.data.flushInput()
    
    def Read_Sensor_Data(self):
        msgs_len = 0x02
        tx_buff = [self.Header, self.Header, msgs_len, self.instruction.Read, self.item.Sensor_Data]
        self.data.write(tx_buff)

        rx_buff = self.data.read(6)
        if len(rx_buff) == 6:
            if(rx_buff[0] == self.Header and rx_buff[1] == self.Header and rx_buff[3] == 0):
                self.data_length = rx_buff[2]
                if(rx_buff[4] == 0x01):
                    self.SensorA = SensorData.DETECTED
                else:
                    self.SensorA = SensorData.NOT_DETECTED
                if(rx_buff[5] == 0x01):
                    self.SensorB = SensorData.DETECTED
                else:
                    self.SensorB = SensorData.NOT_DETECTED
                self.return_data = copy.deepcopy(rx_buff)

            self.return_data = copy.deepcopy(rx_buff)
            self.data.flushInput()
    
    def Read_NFC_Data(self):
        msgs_len = 0x02
        tx_buff = [self.Header, self.Header, msgs_len, self.instruction.Read, self.item.NFC_Data]
        self.data.write(tx_buff)

        rx_buff = self.data.read(7)
        if len(rx_buff) == 7:
            if(rx_buff[0] == self.Header and rx_buff[1] == self.Header and rx_buff[3] == 0):
                self.data_length = rx_buff[2]
                if(rx_buff[4] == 0x01):
                    self.Tag_position = PositionData.HOME
                elif(rx_buff[4] == 0x02):
                    self.Tag_position = PositionData.ON_STATION
                else:
                    self.Tag_position = PositionData.ON_THE_WAY
                self.Tag_value = (rx_buff[5] << 8) | rx_buff[6]
                self.return_data = copy.deepcopy(rx_buff)

            self.return_data = copy.deepcopy(rx_buff)
            self.data.flushInput()
    
    def Read_Joystick_Data(self):
        msgs_len = 0x02
        tx_buff = [self.Header, self.Header, msgs_len, self.instruction.Read, self.item.Joystick_Data]
        self.data.write(tx_buff)

        rx_buff = self.data.read(6)
        if len(rx_buff) == 6:
            if(rx_buff[0] == self.Header and rx_buff[1] == self.Header and rx_buff[3] == 0):
                self.data_length = rx_buff[2]
                self.Xpos = rx_buff[4]
                self.Ypos = rx_buff[5]

            self.return_data = copy.deepcopy(rx_buff)
            self.data.flushInput()
            
UART_COM = Serial_COM(baudRate, USBPort, timeout)

# SETUP RUNNING
ping_checked = False
set_mode_complete = False
set_speed_complete = False
set_accel_complete = False
set_brake_complete = False
setup_done = False

while True:
    if not ping_checked:
        UART_COM.Send_Ping()
        if UART_COM.return_data[0] == UART_COM.Header and UART_COM.return_data[1] == UART_COM.Header and UART_COM.return_data[3] == 0:
            print(" ".join(f"0x{byte:02X}" for byte in UART_COM.return_data))
            ping_checked = True
            print("Ping Check Complete")
            time.sleep(1)
            break
        else:
            print("Waiting Ping Response...")
            ping_checked = False

while True:
    if not set_mode_complete:
        UART_COM.Set_Running_Mode(UART_COM.running_mode.LF_MODE)
        if UART_COM.return_data[0] == UART_COM.Header and UART_COM.return_data[1] == UART_COM.Header and UART_COM.return_data[3] == 0:
            print(" ".join(f"0x{byte:02X}" for byte in UART_COM.return_data))
            set_mode_complete = True
            print("Set Mode Complete")
            time.sleep(1)
            break
        else:
            print("Waiting Running Mode Response...")
            set_mode_complete = False

while True:
    if not set_speed_complete:
        UART_COM.Set_Base_Speed(100)
        if UART_COM.return_data[0] == UART_COM.Header and UART_COM.return_data[1] == UART_COM.Header and UART_COM.return_data[3] == 0:
            print(" ".join(f"0x{byte:02X}" for byte in UART_COM.return_data))
            set_speed_complete = True
            print("Set Speed Complete")
            time.sleep(1)
            break
        else:
            print("Waiting Running Speed Response...")
            set_speed_complete = False

while True:
    if not set_accel_complete:
        UART_COM.Set_Running_Accel(UART_COM.accel_mode.REGENERATIVE_ACCEL)
        if UART_COM.return_data[0] == UART_COM.Header and UART_COM.return_data[1] == UART_COM.Header and UART_COM.return_data[3] == 0:
            print(" ".join(f"0x{byte:02X}" for byte in UART_COM.return_data))
            set_accel_complete = True
            print("Set Accel Complete")
            time.sleep(1)
            break
        else:
            print("Waiting Accel Mode Response...")
            set_accel_complete = False

while True:
    if not set_brake_complete:
        UART_COM.Set_Running_Brake(UART_COM.brake_mode.REGENERATIVE_BRAKE)
        if UART_COM.return_data[0] == UART_COM.Header and UART_COM.return_data[1] == UART_COM.Header and UART_COM.return_data[3] == 0:
            print(" ".join(f"0x{byte:02X}" for byte in UART_COM.return_data))
            set_brake_complete = True
            print("Set Brake Complete")
            time.sleep(1)
            break
        else:
            print("Waiting Brake Mode Response...")
            set_brake_complete = False

while True:
    if ping_checked and set_mode_complete and set_speed_complete and set_brake_complete:
        setup_done = True
        print("All Setup Complete")
        time.sleep(1)
        break

    if setup_done:
        break

# TESTING
while True:
    print("Running")